File: go/0036_valid_sudoku.go
package main

//time: O(1)
//space O(1)

func isValidSudoku(board [][]byte) bool {
	var rows, cols, squares [9]map[byte]bool

	for i := 0; i < 9; i++ {
		rows[i] = map[byte]bool{}
		cols[i] = map[byte]bool{}
		squares[i] = map[byte]bool{}
	}

	for r := 0; r < 9; r++ {
		for c := 0; c < 9; c++ {
			if board[r][c] == '.' {
				continue
			}
			field := board[r][c]
			square := r/3*3 + c/3
			if rows[r][field] || cols[c][field] || squares[square][field] {
				return false
			}
			rows[r][field] = true
			cols[c][field] = true
			squares[square][field] = true
		}
	}

	return true
}

File: csharp/0036-valid-sudoku.cs
// time: O(1)
// space: O(1)

public class Solution {
    public bool IsValidSudoku(char[][] board) {
        var rows = new Dictionary<byte, bool>[9];
        var cols = new Dictionary<byte, bool>[9];
        var squares = new Dictionary<byte, bool>[9];

        for (var i = 0; i < 9; i++)
        {
            rows[i] = new Dictionary<byte, bool>();
            cols[i] = new Dictionary<byte, bool>();
            squares[i] = new Dictionary<byte, bool>();
        }

        for (var r = 0; r < 9; r++)
        {
            for (var c = 0; c < 9; c++)
            {
                if (board[r][c] == '.')
                {
                    continue;
                }
				var field = (byte)(board[r][c] - '0');
                var square = r / 3 * 3 + c / 3;
                if (rows[r].ContainsKey(field) || cols[c].ContainsKey(field) || squares[square].ContainsKey(field))
                {
                    return false;
                }
                rows[r][field] = true;
                cols[c][field] = true;
                squares[square][field] = true;
            }
        }

        return true;
    }
}
